`apiclient apply` support for S3, SSM, and Secrets Manager URIs

anishcheraku opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:** 4208

Closes [#4208](https://github.com/bottlerocket-os/bottlerocket/issues/4208)

**Description of changes:** Refactored the get logic for `apiclient apply`, implemented Structs for each of the resolver types, and added support for S3, SSM, Secrets Manager URIs as well as SSM and Secrets Manager ARNs to `apiclient apply`



**Testing done:** Manual Integration Testing for fetching and modifying configuration data from all sources (Stdin, File, HTTP/HTTPS, S3, SSM, Secrets Manager). Added Unit Tests to verify that the correct resolver is selected, positive and negative test cases to verify that URIs/ARNs are parsed correctly/throw an error, test cases to verify that S3 URIs with special characters and special URIs will be parsed correctly, and added a Unit Test to verify that files are parsed properly using Tempfile.


-- Added by KCSesh:

## Test 1: stdin (-)

<details>

### apiclient apply - TOML from stdin
**Input:**
```bash
echo -e '[settings]\nmotd = "hello from stdin"' | apiclient apply -
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from stdin"
  }
}
```
**Result:** PASS

### apiclient apply - JSON from stdin
**Input:**
```bash
echo '{"settings":{"motd":"stdin json"}}' | apiclient apply -
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "stdin json"
  }
}
```
**Result:** PASS

### apiclient network configure - from stdin
**Input:**
```bash
echo 'version = 2

[eth0]
dhcp4 = true' | apiclient network configure -
cat /.bottlerocket/net.toml
```
**Output:**
```toml
version = 2

[eth0]
dhcp4 = true
```
**Result:** PASS

</details>

---

## Test 2: base64:

<details>

### apiclient apply - base64 encoded TOML
**Input:**
```bash
apiclient apply 'base64:W3NldHRpbmdzXQptb3RkID0gImhlbGxvIGZyb20gYmFzZTY0Igo='
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from base64"
  }
}
```
**Result:** PASS

### apiclient apply - base64 encoded JSON
**Input:**
```bash
apiclient apply 'base64:eyJzZXR0aW5ncyI6eyJtb3RkIjoiYmFzZTY0IGpzb24ifX0='
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "base64 json"
  }
}
```
**Result:** PASS

### apiclient network configure - base64 encoded
**Input:**
```bash
apiclient network configure 'base64:dmVyc2lvbiA9IDIKCltldGgwXQpkaGNwNCA9IHRydWUK'
cat /.bottlerocket/net.toml
```
**Output:**
```toml
version = 2

[eth0]
dhcp4 = true
```
**Result:** PASS

</details>

---

## Test 3: file://

<details>

### apiclient apply - file URI (TOML)
**Input:**
```bash
cat > /tmp/test-settings.toml << 'EOF'
[settings]
motd = "hello from file"
EOF
apiclient apply file:///tmp/test-settings.toml
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from file"
  }
}
```
**Result:** PASS

### apiclient apply - file URI (JSON)
**Input:**
```bash
cat > /tmp/test-settings.json << 'EOF'
{"settings":{"motd":"hello from file json"}}
EOF
apiclient apply file:///tmp/test-settings.json
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from file json"
  }
}
```
**Result:** PASS

### apiclient network configure - file URI
**Input:**
```bash
cat > /tmp/test-net.toml << 'EOF'
version = 3

[eth0]
dhcp4 = true
EOF
apiclient network configure file:///tmp/test-net.toml
cat /.bottlerocket/net.toml
```
**Output:**
```toml
version = 3

[eth0]
dhcp4 = true
```
**Result:** PASS

</details>

---

## Test 4: https://

<details>

### apiclient apply - HTTPS (TOML)
**Input:**
```bash
apiclient apply 'https://gist.githubusercontent.com/KCSesh/6986709bd135c719df9fbce377e84df6/raw/5109180df7fafdfd7f90996dd86750d2988dd6f2/test-settings.toml'
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "Hello from https gist."
  }
}
```
**Result:** PASS

### apiclient apply - HTTPS (JSON)
**Input:**
```bash
apiclient apply 'https://gist.githubusercontent.com/KCSesh/6986709bd135c719df9fbce377e84df6/raw/5109180df7fafdfd7f90996dd86750d2988dd6f2/test-settings.json'
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from https json"
  }
}
```
**Result:** PASS

### apiclient network configure - HTTPS
**Input:**
```bash
apiclient network configure 'https://gist.githubusercontent.com/KCSesh/6986709bd135c719df9fbce377e84df6/raw/5109180df7fafdfd7f90996dd86750d2988dd6f2/test-net.toml'
cat /.bottlerocket/net.toml
```
**Output:**
```toml
version = 2

[eth0]
dhcp4 = true
```
**Result:** PASS

</details>

---

## Test 5: s3://

<details>

### apiclient apply - S3 (TOML)
**Input:**
```bash
apiclient apply "s3://${BUCKET_NAME}/settings.toml"
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from toml"
  }
}
```
**Result:** PASS

### apiclient apply - S3 (JSON)
**Input:**
```bash
apiclient apply "s3://${BUCKET_NAME}/settings.json"
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from json"
  }
}
```
**Result:** PASS

### apiclient network configure - S3
**Input:**
```bash
apiclient network configure "s3://${BUCKET_NAME}/net.toml"
cat /.bottlerocket/net.toml
```
**Output:**
```toml
version = 2

[eth0]
dhcp4 = true
```
**Result:** PASS

</details>

---

## Test 6: ssm://

<details>

### apiclient apply - SSM parameter
**Input:**
```bash
apiclient apply "ssm://${SSM_PARAM}"
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from ssm"
  }
}
```
**Result:** PASS

### apiclient apply - SSM SecureString
**Input:**
```bash
apiclient apply "ssm:///${PREFIX}/secure-settings"
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from ssm secure"
  }
}
```
**Result:** PASS

### apiclient network configure - SSM
**Input:**
```bash
apiclient network configure "ssm:///${PREFIX}/net-config"
cat /.bottlerocket/net.toml
```
**Output:**
```toml
version = 2

[eth0]
dhcp4 = true
```
**Result:** PASS

</details>

---

## Test 7: SSM ARN

<details>

### apiclient apply - SSM ARN
**Input:**
```bash
apiclient apply "${SSM_ARN}"
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from ssm"
  }
}
```
**Result:** PASS

### apiclient network configure - SSM ARN
**Input:**
```bash
apiclient network configure "arn:aws:ssm:<region>:<account-id>:parameter/<parameter-path>"
cat /.bottlerocket/net.toml
```
**Output:**
```toml
version = 2

[eth0]
dhcp4 = true
```
**Result:** PASS

</details>

---

## Test 8: secretsmanager://

<details>

### apiclient apply - Secrets Manager
**Input:**
```bash
apiclient apply "secretsmanager://${SECRET_NAME}"
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from secrets manager"
  }
}
```
**Result:** PASS

### apiclient network configure - Secrets Manager
**Input:**
```bash
apiclient network configure "secretsmanager://${PREFIX}/net-config"
cat /.bottlerocket/net.toml
```
**Output:**
```toml
version = 2

[eth0]
dhcp4 = true
```
**Result:** PASS

### apiclient apply - Secrets Manager ARN
**Input:**
```bash
apiclient apply "${SECRET_ARN}"
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from secrets manager"
  }
}
```
**Result:** PASS

</details>

---

## Test 9: Cross-region SSM ARN (us-east-1)

<details>

### apiclient apply - cross-region SSM ARN
**Input:**
```bash
apiclient apply "arn:aws:ssm:<region>:<account-id>:parameter/<parameter-path>"
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from ssm us-east-1"
  }
}
```
**Result:** PASS

### apiclient network configure - cross-region SSM ARN
**Input:**
```bash
apiclient network configure "arn:aws:ssm:<region>:<account-id>:parameter/<parameter-path>"
cat /.bottlerocket/net.toml
```
**Output:**
```toml
version = 2

[eth0]
dhcp4 = true
```
**Result:** PASS

</details>

---

## Test 10: Cross-region Secrets Manager ARN (us-east-1)

<details>

### apiclient apply - cross-region Secrets Manager ARN
**Input:**
```bash
apiclient apply "arn:aws:secretsmanager:<region>:<account-id>:secret:<secret-name>"
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from us-east-1"
  }
}
```
**Result:** PASS

### apiclient network configure - cross-region Secrets Manager ARN
**Input:**
```bash
apiclient network configure "arn:aws:secretsmanager:<region>:<account-id>:secret:<secret-name>"
cat /.bottlerocket/net.toml
```
**Output:**
```toml
version = 2

[eth0]
dhcp4 = true
```
**Result:** PASS

</details>

---

## Test 11: Multiple sources

<details>

### apiclient apply - multiple sources in one command
**Input:**
```bash
apiclient apply \
  "s3://${BUCKET_NAME}/settings.toml" \
  "ssm://${SSM_PARAM}" \
  'base64:W3NldHRpbmdzXQptb3RkID0gImhlbGxvIGZyb20gYmFzZTY0Igo='
apiclient get settings.motd
```
**Output:**
```json
{
  "settings": {
    "motd": "hello from base64"
  }
}
```
**Result:** PASS (last source wins for same key)

</details>

---

## Test 12: Error cases

<details> 

### Invalid base64

**Input:**
```
apiclient apply 'base64:not-valid-base64!!!'
```

**Output:**
```
Failed to apply settings: Resolver failed: Failed to decode base64 data from 'base64:not-valid-base64!!!': Invalid symbol 45, offset 3.
```

**Result:** PASS (clear error message)

---

### Non-existent S3 object

**Input:**
```
apiclient apply 's3://nonexistent-kcsesh-bucket-12345/file.toml'
```

**Output:**
```
Failed to apply settings: Resolver failed: TLS resolver error: Failed to fetch S3 object 'nonexistent-kcsesh-bucket-12345/file.toml': NoSuchBucket: The specified bucket does not exist
```

**Result:** PASS (clear error message)

---

### Non-existent SSM parameter

**Input:**
```
apiclient apply 'ssm:///nonexistent/parameter'
```

**Output:**
```
Failed to apply settings: Resolver failed: TLS resolver error: Failed to fetch parameter '/nonexistent/parameter' from SSM: ParameterNotFound
```

**Result:** PASS (clear error message)

---

### Non-existent secret

**Input:**
```
apiclient apply 'secretsmanager://nonexistent-secret'
```

**Output:**
```
Failed to apply settings: Resolver failed: TLS resolver error: Failed to fetch secret 'nonexistent-secret' from Secrets Manager: AccessDeniedException: User: arn:aws:sts::XXXXXXXXXXXX:assumed-role/BottlerocketCoreKitTest/i-00fb94b18224a962d is not authorized to perform: secretsmanager:GetSecretValue on resource: nonexistent-secret because no identity-based policy allows the secretsmanager:GetSecretValue action
```

**Result:** PASS (clear error message)

---

### Invalid URI scheme

**Input:**
```
apiclient apply 'ftp://example.com/file.toml'
```

**Output:**
```
Failed to apply settings: Resolver failed: No URI resolver found for 'ftp://example.com/file.toml'
```

**Result:** PASS (clear error message)

---

### Malformed ARN

**Input:**
```
apiclient apply 'arn:aws:invalid'
```

**Output:**
```
Failed to apply settings: Resolver failed: No URI resolver found for 'arn:aws:invalid'
```

**Result:** PASS (clear error message)
</details>

## Test 13: Max Size

<details>

* Verified that a 2MiB file correctly applies. 

```
[ssm-user@control]$ apiclient apply s3://<>-bucket/stress-test-containers-2mb-exact.toml
Works!
```


* Verified that a 2MiB file + 1 byte does not apply. 
```
[ssm-user@control]$ apiclient apply s3://<>-bucket/stress-test-containers-2mb-over.toml
Failed to apply settings: Resolver failed: TLS resolver error: S3 object s3://<>-bucket/stress-test-containers-2mb-over.toml is too large (2097153 bytes, maximum is 2097152 bytes)
[ssm-user@control]$

```

* Verified a loop of 200 iteration of apiclient apply s3://<>-bucket/stress-test-containers-2mb-exact.toml  


</details>



## Test 14: Different Partition

<details>

* Verified that ARN parameters apply on `aws-us-gov` for S3/SSM/Secrets manager

</details>

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
apiclient: apply: refactor get logic

Co-authored-by: Kyle Sessions <kssessio@amazon.com>

Commit:
apiclient: apply: add support for s3 and secretmanger URIs

Co-authored-by: Kyle Sessions <kssessio@amazon.com>

anishcheraku commented:
Refactored imports based on code review feedback, and improved commit messages.

jmt-lab commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,159 @@
+use reqwest::Url;
+use crate::apply::{Error, Result};
+use crate::apply::error::{
+    FileReadSnafu, FileUriSnafu, ReqwestSnafu, StdinReadSnafu, S3UriMissingBucketSnafu, InvalidFileUriSnafu, InvalidHTTPUriSnafu, S3UriSchemeSnafu
```
Why do we have both FileUriSnafu and InvalidFileUriSnafu
which is a  bit confusing? Can we differentiate these error variants?

anishcheraku replied:
Fixed - Got rid of `InvalidFileUriSnafu` because it did the same thing as `FileUriSnafu` and to maintain consistency with other resolvers.

jmt-lab commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,159 @@
+            .await
+            .context(ReqwestSnafu {
+                uri: self.url.to_string(),
+                method: "GET".to_string(),
```
You don't need .to_string() here

anishcheraku replied:
Fixed - Addressing this in another comment: https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/554#discussion_r2178409960

jmt-lab commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,159 @@
+#[async_trait]
+impl UriResolver for S3Uri {
+    async fn resolve(&self) -> Result<String> {
+        // still unimplemented
```
We generally don't want to merge a PR with unimplemented methods like this, but since this is a draft I guess its alright for now

cbgbt replied:
Agree with Jarrett, but as another tip, in rust you can use the `unimplemented!()` macro in functions to avoid having to "make stuff up" Like this:

```rust
async fn resolve(&self) -> Result<String> {
    unimplemented!()
}
```

Instead of having to make up an `Err` to return, `unimplemented!()` just does what it takes to make the compiler happy (it panics, lol)

anishcheraku replied:
Fixed - I've now implemented the resolve() method completely, so I have removed the unimplemented comment.

jmt-lab commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,159 @@
+// src/uri_resolver.rs
```
This module is lacking alot of documentation, especially a module level devinition, maybe addint something like:

```
//! URI resolution for different protocols.
//!
//! This module provides a trait-based approach to resolving different types of URIs
//! including stdin (`-`), file paths (`file://`), HTTP(S) URLs, and S3 URIs.
```

cbgbt replied:
Try to imagine what someone reading this file for the first time would want.

The module-level docstring is your opportunity to explain how the abstractions/types in here relate to eachother.

If you prime programmers with the right information here, they will be able to read and comprehend your code *much* faster.

With more sparse information, you basically make the programmer guess at how the implementation works. Then they have to read the code to be absolutely certain (or what often times happen is that they don't do this, and then are surprised when their mental model doesn't totally match the actual code)

anishcheraku replied:
Fixed - Added a module-level docstring to the top of `uri_resolver.rs`

jmt-lab commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -67,46 +71,40 @@ where
-            .context(error::StdinReadSnafu)
-            .await?;
-        return Ok(output);
+pub async fn get(input: &str) -> Result<String> {
```
Is there a reason the visibility on this method needs to be bumped to pub?

anishcheraku replied:
Fixed - the function is no longer pub. It was temporarily bumped to `pub` in the PR because one of the variables didn't have the correct scope.

jmt-lab commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -67,46 +71,40 @@ where
+    let url = Url::parse(input).context(error::UriSnafu { input_source: input.to_string() })?;
+
+    // 3) file://
+    if let Ok(r) = FileUri::try_from(url.clone()) {
```
I wonder if it might be better to use a prefix strip and then match instead of trying to directly form the Uri type directly and failing out, but its not a real blocker here.

anishcheraku replied:
Fixed - fixed this issue in another comment: https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/554#discussion_r2178622051

anishcheraku commented:
Pushed new Resolver for `secretsmanager://` and added error handling for Secrets Manager. (reverted in next commit)

anishcheraku commented:
Reversed the changes I just made for `secretsmanager://`

anishcheraku commented:
Pushed new Resolver for secretsmanager:// and added error handling for Secrets Manager.

jmt-lab reviewed (commented):
Looks like something went awry with the rebase packages/kubernetes-1.26 files are showing up here.

jmt-lab commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -83,11 +84,16 @@ fn select_resolver(input: &str) -> Result<Box<dyn UriResolver>> {
         return Ok(Box::new(r));
     }
 
+    // 6) secretsmanager://
```
These numberings look out of order, I used to do numbering too but now i've learned better to not do numbering as it prevents edits from creating inconsistencies.

anishcheraku replied:
Fixed - Removed the numberings and fixed the order.

cbgbt commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -67,46 +72,45 @@ where
+}
+
+/// Choose which UriResolver applies to `input` (stdin, file://, http(s):// or s3://).
+fn select_resolver(input: &str) -> Result<Box<dyn UriResolver>> {
```
This function is fine, but I somewhat prefer an implementation where we have a static list of resolvers, and the resolver selection iterates through the list, calling an interface that each resolver exposes to decide if that resolver is capable of handling this input.

This makes it a little easier to reason about how to add more resolvers, rather than having to keep the stateful "if/else" ladder in your mind.


```
trait ResolverSelector {
    fn try_create_resolver(input: &str) -> Option<Box<dyn UriResolver>>;
}

const RESOLVERS: &[&dyn ResolverSelector] = &[
    &StdioResolverSelector,
    &FileResolverSelector,
    // etc
];

struct FileResolverSelector;

impl ResolverSelector for FileResolverSelector {
    fn try_create_resolver(input: &str) -> Option<Box<dyn UriResolver>> {
        let input_uri = Url::parse(&input).ok()?; // This maps the "Result" response into an "Option"
        // try to make a FileResolver or return None if we can't
    }
}
```

KCSesh replied:
As I understand this - it is now resolved

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,208 @@
+
+use reqwest::Url;
+use crate::apply::{Error, Result};
+use crate::apply::error::{
```
We shouldn't import Snafu's like this.

The "preferred" way is, in each function that needs these Snafu context selectors, do something like this:


```rust
fn something_that_can_fail() -> Result<(), ErrorType> {
    use crate::apply::error::*;
    // now the selectors are in scope just for this function
}
```

cbgbt replied:
One reason is that this import list basically becomes a chore to maintain.

anishcheraku replied:
Fixed - Used the preferred format for `crate::uri_resolver` as well as `crate::apply`.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,208 @@
+                method: "GET".to_string(),
+            })?
+            .error_for_status()
+            .context(ReqwestSnafu {
```
These shouldn't all be `ReqwestSnafu`. We want a different error variant to explain the failure mode more succinctly (e.g. an error variant for failing to get `text()` from a response body should be different than the error variant for failing to make a web request at all)

anishcheraku replied:
Fixed - replaced `ReqwestSnafu`s with `HttpRequest`, `HttpStatus`, and `HttpBody`.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,208 @@
+    }
+}
+
+/// s3://bucket/key URLs (stub; add aws-sdk-s3 later)
```
This is useful information to include in a docstring, but you should start with a human-readable description first. Something like:


```
/// Uri Resolver that fetches content from S3
///
/// This resolver accepts input of the form s3://bucket/key and translates this into
/// authenticated AWS S3 get requests using the standard AWS credential resolution mechanism

anishcheraku replied:
Fixed - added a description for each resolver.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,208 @@
+        use aws_config::{self, BehaviorVersion, Region};
+        use aws_sdk_secretsmanager;
+
+        // 1) load AWS config (region/account via env)
```
In general, comments which exactly say what the code below does aren't a great idea. The newline separation you have in this function is good, and you can basically skim each newline-separated section to figure out the flow of the function.

I think a better approach would be to hoist a lot of this information into the docstring:


```rust

/// Resolves incoming source artifacts by fetching from AWS Secrets Manager
///
/// This function uses the typical AWS credentials resolution mechanism to authenticate.
```

Docstrings like this are useful at two levels:
* People calling this function don't need to read the code to understand what's going to happen
* People who need to change this code can just read the docstring and be primed to understand what they're about to read, much like how I mentioned with module-level docstrings.


---

So when is it more useful to have inner comments?
* If the discrete actions of the function are hard to figure out by skimming each newline-separated section (this can also be a sign that you should refactor things, but like all things it's a tradeoff decision)
* If the *what* of the code is clear, but not the *why* (e.g. Imagine a function much like this one, but it reads AWS configurations from a separate file. You would probably need to add a comment there explaining why you are doing that)

cbgbt reviewed (commented):
Nice work!

sam-berning commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -1,9 +1,10 @@
-
-use aws_sdk_s3::Client;
-use crate::uri_resolver::{StdinUri, FileUri, HttpUri, S3Uri, UriResolver};
+//! use aws_smithy_runtime_api::client::result::SdkError;
```
looks like this commented out code is no longer needed?

anishcheraku replied:
Fixed - removed all unneeded imports

sam-berning commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -51,10 +53,10 @@ pub struct FileUri {
 }
 
-impl TryFrom<Url> for FileUri {
+impl TryFrom<&Url> for FileUri {
```
I think we should standardize on one type for TryFrom impls. it's a little strange that each of these Uri types comes from a different type. once you've done your testing on which approach handles special characters better, we can align on one type to use for all of these TryFrom impls.

assuming that both implementations behave the same way, I think I prefer `TryFrom<Url>` and using `url.scheme()` for matching over `TryFrom<&str>` and manual parsing.

anishcheraku replied:
Fixed - all resolvers are now using url.scheme() to match.

cbgbt commented on Twoliter.toml:
```diff
@@ -1,5 +1,5 @@
 schema-version = 1
-release-version = "9.1.0"
+release-version = "8.2.0-1"
```
We'll need to make sure we pull these changes out prior to merging.

anishcheraku replied:
Fixed - Rebased my code on top of the `bottlerocket-os:develop` branch and stopped tracking changes to Twoliter.toml.

cbgbt commented on sources/api/apiclient/Cargo.toml:
```diff
@@ -31,7 +37,7 @@ log.workspace = true
 nix.workspace = true
 rand = { workspace = true, features = ["default"] }
-reqwest.workspace = true
+reqwest.workspace = true 
```
nit:

```suggestion
reqwest.workspace = true
```

anishcheraku replied:
Fixed - Removed the extra space

cbgbt commented on sources/api/apiclient/Cargo.toml:
```diff
@@ -44,6 +50,7 @@ tokio-tungstenite = { workspace = true, features = ["connect"] }
 toml.workspace = true
 unindent.workspace = true
 url.workspace = true
+aws-smithy-runtime-api.workspace = true
```
I don't think we actually want to directly take a dependency here. Most SDK crates re-export the error types they create.

anishcheraku replied:
Fixed - This dependency is no longer being used

cbgbt replied:
It appears to still be present in `Cargo.toml`

cbgbt commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -69,44 +70,52 @@ where
+}
+
+
+/// Choose which UriResolver applies to `input` (stdin, file://, http(s):// or s3://).
```
This appears to be out-of-date.

anishcheraku replied:
Fixed - Updated the comment to reflect the new resolvers that were added.

cbgbt commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -218,9 +237,77 @@ mod error {
         #[snafu(display("Failed to read standard input: {}", source))]
         StdinRead { source: std::io::Error },
 
+        #[snafu(display("Failed to read S3 object body '{bucket}/{key}': {}", source))]
+        S3Body {
+            bucket: String,
+            key: String,
+            source: aws_sdk_s3::primitives::ByteStreamError,
+        },
+
+        #[snafu(display("Failed to fetch S3 object '{bucket}/{key}': {}", source))]
+        S3Get {
+            bucket: String,
+            key: String,
+            source: SdkError<GetObjectError, SdkHttpResponse>,
+        },
```
Instead of expanding the top-level error type with all of these error variants, I think it would be better to instead make a ResolverError type in the `uri_resolver` module.

We could then add `Resolver` as a specific type here. This would make it easier to determine at what point in the program each error variant can occur.

anishcheraku replied:
Fixed - Moved all error variants associated with the resolvers into `uri_resolver.rs` into a new `ResolverError` enum. Additionally, the functions now return a new type `ResolverResult`, so it is easier to associate errors with their sources.

cbgbt commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -142,11 +151,18 @@ fn format_change(input: &str, input_source: &str) -> Result<String> {
+    use aws_sdk_secretsmanager::operation::get_secret_value::GetSecretValueError;
+    use aws_sdk_ssm::operation::get_parameter::GetParameterError;
+    use crate::apply::SdkHttpResponse;
+    use aws_smithy_runtime_api::client::result::SdkError;
```
We probably don't want this.

anishcheraku replied:
Resolved - The S3Get error variant is now getting the SdkError from the `aws_sdk_s3` crate instead of the `aws_smithy_runtime_api` crate. The SdkError for both SSM and Secrets Manager were already using their respective crates.

cbgbt commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -142,11 +151,18 @@ fn format_change(input: &str, input_source: &str) -> Result<String> {
+    use crate::apply::SdkHttpResponse;
+    use aws_smithy_runtime_api::client::result::SdkError;
+    use aws_sdk_s3::operation::get_object::GetObjectError;
+    //use aws_smithy_http::result::SdkHttpResponse;
```
Please remove unused code.

anishcheraku replied:
Fixed - I have removed the unused code

cbgbt commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -69,44 +70,52 @@ where
+    let url = Url::parse(input).context(error::UriSnafu { input_source: input.to_string() })?;
+
+    // file://
+    if let Ok(r) = crate::uri_resolver::FileUri::try_from(&url) {
```
I might suggest adding `use crate::uri_resolver` so we don't have to fully specify `crate::` each time here.

anishcheraku replied:
Fixed - Imported the `uri_resolver` crate, so that I no longer have to specify `crate::` each time

cbgbt commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -69,44 +70,52 @@ where
+
+
+/// Choose which UriResolver applies to `input` (stdin, file://, http(s):// or s3://).
+fn select_resolver(input: &str) -> Result<Box<dyn crate::uri_resolver::UriResolver>> {
```
Can we add unit tests to assert that the correct resolver is selected?

I also think we really need tests to understand what happens when characters in the S3 keyspace are used which aren't supported by [URL standard](https://url.spec.whatwg.org/). We should add tests for these cases, as it seems like it's entirely up to our parser implementation to decide what to do given this current implementation.

anishcheraku replied:
Fixed - I have added unit tests to assert that the correct resolver was selected. 

Fixed - After trying out 4 different try_from() implementations for the S3Uri, I have found one that works on all sorts of URIs that have special characters, and ones that don't follow the URL standard. I have also added unit tests for testing different S3 URIs that have special characters and don't follow the URL standard.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,319 @@
+    parameter_name: String,
+}
+
+impl TryFrom<&str> for SsmUri {
```
I think we're still supporting `TryFrom` on multiple types. Does it make sense to unify on each resolver accepting `&str` types and allowing them each to parse?

cbgbt replied:
This is technically less efficient than the current setup, but it would make it easier for us to test our Resolver selector.

Another option could be to do something like this:

```rust

struct SettingsInput {
    input: String,
    parsed_url: Option<Url>,
}

impl SettingsInput {
    pub(crate) fn new(input: impl Into<String>) -> Self {
        let input = input.into();
        let parsed_url = Url::parse(&input).ok();
        Self { input, parsed_url }
    }
}

```

Then you'd change `select_resolver` to be like this:

```rust
fn select_resolver(input: &SettingsInput) -> Result<Box<dyn crate::uri_resolver::UriResolver>> {
    // ...
}
```

And you could make it so that each resolver had a uniform `TryFrom` impl on `&SettingsInput`.

Doing this would let you parse the Url exactly once and standardize the `TryFrom`. Resolvers that require a Url could basically return an error if the `parsed_url` is `None`.

anishcheraku replied:
Fixed - `TryFrom` now takes `&SettingsInput` as its input, and determines what input type is required for each resolver. Also, only one type (`&SettingsInput`) is passed into all of the resolvers.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,319 @@
+/// This resolver accepts URIs of the form `file:///path/to/file` (or on Windows
+/// `file://C:/path/to/file`), converts them to a `PathBuf`, and returns the
+/// file’s entire contents as a UTF-8 string.
+pub struct FileUri {
```
We should probably add unit tests for this one using tempfile.

anishcheraku replied:
Fixed - added a unit test that created and wrote to a tempfile, generated a file:// URI, passed it, and asserted that it returns the correct contents.

cbgbt commented:
I think the build failed because the `Cargo.lock` file is not in-sync with your dependency closure?

You can fix this locally by going into the `sources` directory and doing `cargo check`.

sam-berning commented on sources/api/apiclient/Cargo.toml:
```diff
@@ -39,11 +44,13 @@ serde_json.workspace = true
 signal-hook.workspace = true
 simplelog.workspace = true
 snafu = { workspace = true, features = ["futures"] }
+test-case = "*"
```
please use the test-case dependency in the workspace:

```suggestion
test-case.workspace = true
```

anishcheraku replied:
Fixed - switched dependencies

sam-berning commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -142,11 +162,12 @@ fn format_change(input: &str, input_source: &str) -> Result<String> {
 }
 
-mod error {
+pub(crate) mod error {
```
does this need to be `pub` anymore?

anishcheraku replied:
Fixed - it is no longer required for the `error module` to be visible within the entire crate since the error handling for the resolvers have been moved to `uri_resolver.rs`

sam-berning commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -142,11 +162,12 @@ fn format_change(input: &str, input_source: &str) -> Result<String> {
 
     #[derive(Debug, Snafu)]
-    #[snafu(visibility(pub(super)))]
+    #[snafu(visibility(pub(crate)))]
```
same here, this should only be needed in apply, right?

anishcheraku replied:
Fixed - We no longer need it to be visible to the entire crate.

sam-berning commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -218,8 +233,8 @@ mod error {
-        #[snafu(display("Failed to read standard input: {}", source))]
-        StdinRead { source: std::io::Error },
+        #[snafu(display("Given invalid file URI '{}'", input_source))]
+        InvalidFileUri { input_source: String },
```
nit: this seems unused?

anishcheraku replied:
Fixed - This error variant is no longer required. There is another variant (`FileUri`) in the `ResolverError` enum for handling invalid file URIs.

sam-berning commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,414 @@
+            .parsed_url
+            .clone()
+            .context(FileUriSnafu { input_source: input.input.clone() })?;
+        use resolver_error::*;
```
nit: try to put `use` statements as early in the code as you can

(does this compile? seems like `FileUriSnafu` on line 71 should be an unknown type)

anishcheraku replied:
Fixed - Moved the `use statement` to the top of the `try_from` method. Also, the program compiles and the error variant works as intended.

sam-berning replied:
can you do this for all of the `try_from` methods please?

anishcheraku replied:
Fixed - I forgot to move one use statement to the top, but it is fixed now.

sam-berning commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,414 @@
+        // must start with "secretsmanager://"
+        ensure!(
+            input.input.as_str().starts_with("secretsmanager://"),
+            SecretsManagerUriSnafu { input_source: input.input.as_str().to_string() }
```
nit: `input.input.clone()` would work here

anishcheraku replied:
Fixed - replaced `.as_str().to_string()` with `.clone()`.

sam-berning replied:
you do this in the SsmUri TryFrom<&SettingsInput> implementation as well, could be good to update there too

anishcheraku replied:
Fixed - all statements only use one of the following: `.clone()`, `.to_string()`, or `.as_str()`.

sam-berning commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,414 @@
+            SecretsManagerUriSnafu { input_source: input.input.as_str().to_string() }
+        );
+        // strip the prefix and ensure there's actually an ID
+        let id = &input.input.as_str()["secretsmanager://".len()..];
```
nit: you could use `.strip_prefix("secretsmanager://")` here, which would be easier to read

anishcheraku replied:
Fixed - replaced `["secretsmanager://".len()..]` with `.strip_prefix("secretsmanager://")`. Also rewrote some of the code in the `try_from()` method to account for changing the type of `id`.

sam-berning replied:
let's update this in the TryFrom implementation for SsmUri as well. I think the S3 resolver's TryFrom implementation is the cleanest of the AWS resolvers as well, can you have all three of them align with that implementation?

anishcheraku replied:
Fixed - All `TryFrom` implementations are significantly cleaner and parsing logic for `SSM and Secrets Manager` align similarly with the parsing logic in S3Uri's TryFrom implementation.

sam-berning commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,414 @@
+        source: std::io::Error,
+    },
+
+    #[snafu(display("Given HTTP(S) URI '{}'", input_source))]
```
```suggestion
    #[snafu(display("Given invalid HTTP(S) URI '{}'", input_source))]
```

anishcheraku replied:
Fixed - fixed the grammatical mistake

sam-berning commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,414 @@
+    S3Get {
+        bucket: String,
+        key: String,
+        source: aws_sdk_s3::error::SdkError<aws_sdk_s3::operation::get_object::GetObjectError, aws_sdk_secretsmanager::config::http::HttpResponse>,
```
```suggestion
        source: aws_sdk_s3::error::SdkError<aws_sdk_s3::operation::get_object::GetObjectError, aws_sdk_s3::config::http::HttpResponse>,
```

anishcheraku replied:
Fixed - the `S3Get` error variant is now using the right sdk for its error source.

sam-berning commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,414 @@
+use std::path::PathBuf;
+use tokio::io::AsyncReadExt;
+use reqwest::Url;
+pub type ResolverResult<T> = std::result::Result<T, ResolverError>;
```
nit: don't mix type declarations in use statements. I typically put a `Result` alias at the bottom of a file, but that's personal preference

anishcheraku replied:
Fixed - moved `ResolverResult<T>` to the bottom of the file.

Commit:
apiclient: apply: add support for SSM and Secrets Manager ARNs

Co-authored-by: Kyle Sessions <kssessio@amazon.com>

sam-berning commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,846 @@
+impl UriResolver for HttpUri {
+    async fn resolve(&self) -> ResolverResult<String> {
+        use resolver_error::*;
+        const MAX_SIZE_BYTES: u64 = 100 * 1024 * 1024;
```
You should share this value between the two resolvers that use it, so we can configure it in one place

anishcheraku replied:
Fixed - `MAX_SIZE_BYTE` is now defined at the top of the `uri_resolver` crate, and has a comment explaining its purpose.

sam-berning commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,846 @@
+impl TryFrom<&SettingsInput> for SecretsManagerArn {
+    type Error = ResolverError;
+
+    fn try_from(input: &SettingsInput) -> ResolverResult<Self> {
```
this logic is really similar to the `SsmArn::try_from()` logic, consider creating a high level Arn type that handles parsing the Arn and reusing it in each of these resolvers to reduce code duplication

cbgbt commented on sources/api/apiclient/Cargo.toml:
```diff
@@ -39,11 +45,14 @@ serde_json.workspace = true
 signal-hook.workspace = true
 simplelog.workspace = true
 snafu = { workspace = true, features = ["futures"] }
+test-case.workspace = true
```
This should be listed under `[dev-dependencies]` instead of `[dependencies]`

anishcheraku replied:
Fixed - moved `test-case.workspace` to `[dev-dependencies]`.

cbgbt commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -66,47 +66,77 @@ where
+impl SettingsInput {
+    pub(crate) fn new(input: impl Into<String>) -> Self {
+        let input = input.into();
+        let parsed_url = Url::parse(&input).ok();
+        SettingsInput { input, parsed_url }
```
Wouldn't be a bad idea to emit a `debug!()` message if the URL parse fails that includes the error message from that failure. That would help us debug cases where the URL parser fails when we would expect it to succeed.

anishcheraku replied:
Fixed - a `log::debug!` message is now logged whenever the File or HTTP URI resolvers fails to parse a URL correctly.

cbgbt replied:
I'd rather see the debug message logged here where we still have access to the error message that would have been returned by `Url::parse()`.

The error will only be meaningful in the File/HTTP URI cases, but the idea is that the message will be logged at `debug`, allowing it to be ignored in cases where it isn't relevant.

cbgbt commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -66,47 +66,77 @@ where
+    // stdin ("-")
+    if let Ok(r) = uri_resolver::StdinUri::try_from(input) {
+        return Ok(Box::new(r));
     }
```
The repetition in this function is still hard for me to thumbs up. I have a simple suggestion to improve this via macros:

```rust
/// Macro to try multiple settings resolver types in sequence, returning the first one that succeeds.
macro_rules! try_resolvers {
    ($input:expr, $($resolver_type:ty),+ $(,)?) => {
        $(
            if let Ok(r) = <$resolver_type>::try_from($input) {
                return Ok(Box::new(r));
            }
        )+
    };
}

/// Choose which UriResolver applies to `input` (stdin, file://, http(s)://, s3://, secretsmanager://, and ssm://).
fn select_resolver(input: &SettingsInput) -> Result<Box<dyn crate::uri_resolver::UriResolver>> {
    use crate::uri_resolver::*;

    try_resolvers!(
        input,
        StdinUri,
        FileUri,
        HttpUri,
        S3Uri,
        SecretsManagerArn,
        SecretsManagerUri,
        SsmArn,
        SsmUri,
    );

    error::NoResolverSnafu {
        input_source: input.input.clone(),
    }
    .fail()
}
```

cbgbt replied:
We should also probably emit a `debug!`-level log message when we select a resolver saying which one we are using.

anishcheraku replied:
Fixed - Made both changes.

cbgbt commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -218,25 +242,54 @@ mod error {
+    #[test_case("secretsmanager://sec", TypeId::of::<crate::uri_resolver::SecretsManagerUri>(); "secrets")]
+    #[test_case("ssm://param",          TypeId::of::<crate::uri_resolver::SsmUri>();           "ssmUri")]
+    #[test_case("arn:aws:ssm:<region>:<account_id>:parameter/<name>",          TypeId::of::<crate::uri_resolver::SsmArn>();           "ssmArn")]
+    #[test_case("arn:aws:secretsmanager:<region>:<account-id>:secret:<secret-id>",          TypeId::of::<crate::uri_resolver::SecretsManagerArn>();           "secretsmanagerArn")]
```
The spacing on these is very odd. You should either consistently use one space after each comma, or modify these to look like this:

```rust
#[test_case(
    "-",
    TypeId::of::<crate::uri_resolver::StdinUri>()
    ; "stdin")]
```

anishcheraku replied:
Fixed - The spacing is now consistent for all test cases.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,846 @@
+    fn try_from(input: &SettingsInput) -> ResolverResult<Self> {
+        use resolver_error::*;
+
+        let url = input.parsed_url.clone().context(FileUriSnafu {
```
No need to clone.

```suggestion
        let url = input.parsed_url.as_ref().context(FileUriSnafu {
```

anishcheraku replied:
Fixed - Using `as_str()` instead of `clone()`

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,846 @@
+    async fn resolve(&self) -> ResolverResult<String> {
+        use resolver_error::*;
+        const MAX_SIZE_BYTES: u64 = 100 * 1024 * 1024;
+        // 1) issue the GET
```
I'd remove these comments

anishcheraku replied:
Fixed - Removed comments that explained obvious code.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,846 @@
+            })?;
+
+        // 2) check status
+        let resp = if resp.status().is_success() {
+            resp
+        } else {
+            return HttpStatusSnafu {
+                uri: self.url.to_string(),
+                status: resp.status(),
+            }
+            .fail();
+        };
```
`resp` isn't actually changed in this, so it's idiomatically/stylistically better written with `ensure!`

```suggestion
        ensure!(resp.status().is_success(), HttpStatusSnafu {
                uri: self.url.to_string(),
                status: resp.status(),
            }
        );
```

anishcheraku replied:
Fixed - replaced `if, else` with `ensure` since `resp` stays the same.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,846 @@
+                    max_size: MAX_SIZE_BYTES,
+                    uri: self.url.to_string(),
+                });
+            }
```
```suggestion
            ensure!(content_length <= MAX_SIZE_BYTES, HttpObjectTooLargeSnafu {
                size: content_length,
                max_size: MAX_SIZE_BYTES,
                uri: self.url.to_string(),
            });
```

anishcheraku replied:
Fixed - replaced existing code with the `ensure` statement.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,846 @@
+            uri: self.url.to_string(),
+        })?;
+
+        if bytes.len() as u64 > MAX_SIZE_BYTES {
```
Use `ensure!` here as well.

anishcheraku replied:
Fixed - replaced existing code with `ensure!`

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,884 @@
+        })?;
+
+        // 5) UTF-8 decode
+        Ok(String::from_utf8_lossy(&bytes.into_bytes()).into_owned())
```
We shouldn't use `from_utf8_lossy` here and should instead return an error if the decoded content is not utf8. We do this above already.

anishcheraku replied:
Fixed - Used the same logic in the `resolve()` function for `HttpUri`, which now returns an error if the decoded content is not utf8.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,884 @@
+            key: self.key.clone(),
+        })?;
+
+        // 5) UTF-8 decode
```
I would still advocate for removing comments like this, most of them are clear from the code.

anishcheraku replied:
Fixed - Removed comments that explained obvious code.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,884 @@
+            }
+        );
+
+        // split into exactly 6 parts: ["arn","aws",service,region,account,rest]
+        let parts: Vec<&str> = input.split(':').collect();
+        ensure!(
+            parts.len() > 5,
+            InvalidArnFormatSnafu {
```
If we are expecting exactly 6 parts, why not ensure that the length is exactly 6?

The implication seems to be that `rest` can contain `:` characters. Is that true? If so, can we explicitly say that in this comment?

anishcheraku replied:
Fixed - `:` is not a valid character for Secrets Manager secrets or SSM parameters. Both resolvers for `SecretsManagerArn` and `SsmArn` now ensure that `arn.parts` is an exactly some integer.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,884 @@
+impl UriResolver for SecretsManagerArn {
+    async fn resolve(&self) -> ResolverResult<String> {
+        use resolver_error::*;
+        // 1) Load default config, then override region & credentials
```
It seems like we're not overriding the credentials.

anishcheraku replied:
Fixed - Forgot to update the comment. The original comment mentioned overwriting the credentials because that was my first approach, but then decided to use IAM permissions to access parameters and secrets instead (doesn't require overriding credentials).

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,884 @@
+        // 3) Construct
+        Ok(SsmArn {
+            region: arn.region,
+            parameter_name: input.input.to_string(), // still use the full `/parameter/...` segment
```
The comment and `parameter_name` are both incorrect. This contains the full ARN.

anishcheraku replied:
Fixed - Fixed both variable names for SecretsManagerArn and SsmArn to be more descriptive of what they actually contain (`full_arn`).

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,884 @@
+impl UriResolver for SsmArn {
+    async fn resolve(&self) -> ResolverResult<String> {
+        use resolver_error::*;
+        // 1) Load default config, then override region & credentials
```
Comment is not correct, we aren't changing credentials.

anishcheraku replied:
Fixed - Forgot to update the comment. The original comment mentioned overwriting the credentials because that was my first approach, but then decided to use IAM permissions to access parameters and secrets instead (doesn't require overriding credentials).

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,884 @@
+            })?;
+
+        Ok(value)
+    }
+}
+
+#[derive(Debug, Snafu)]
+#[snafu(module)]
+pub enum ResolverError {
+    #[snafu(display("Invalid ARN '{}': {}", input_source, reason))]
+    InvalidArnFormat {
```
I'd prefer having this be composed of smaller Resolver-specific types like `FileUriResolverError`, `S3ResolverError`, `SsmResolverError`, etc would improve maintainability.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,884 @@
+    }
+
+    //FileUri
+    #[test_case("file:///tmp/foo", "/tmp/foo"; "file_ok")]
+    fn parse_file(input: &str, expected_path: &str) {
```
These tests are good, but we aren't doing any negative validation in any of these.

What inputs should cause the FileUri resolver to fail to parse? Same for all of the other ones. I'd like to see test-based proof that we won't parse an SSM ARN as a secretsmanager one.

anishcheraku replied:
Fixed - Added a lot of negative test cases for all of the resolvers.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,884 @@
+    use crate::apply::SettingsInput;
+    use std::convert::TryFrom;
+    use test_case::test_case;
+    #[test_case("s3://testbucket/🚀⚡️.json", "testbucket", "🚀⚡️.json"; "s3_emoji")]
+    #[test_case("s3://testbucket/#hashstart.json", "testbucket", "#hashstart.json"; "hash_start")]
+    #[test_case("s3://testbucket/@atstart.json", "testbucket", "@atstart.json"; "at_start")]
+    #[test_case("s3://testbucket/'singlequotes'.json", "testbucket", "'singlequotes'.json"; "single_quotes")]
+    #[test_case("s3://testbucket/$dollarstart.json", "testbucket", "$dollarstart.json"; "dollar_start")]
+    #[test_case("s3://testbucket/doublejson.json.json", "testbucket", "doublejson.json.json"; "double_dot_json")]
```
Nice tests! Can you include a `?` in one of these as well?

anishcheraku replied:
Fixed - Added a test case with `?`s.

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,992 @@
+    fn try_from(input: &SettingsInput) -> ResolverResult<Self> {
+        use resolver_error::*;
+
+        let url = {
+            if let None = &input.parsed_url {
+                log::debug!("Failed to parse HTTP URI '{}'", input.input);
+            }
+            input.parsed_url.clone()
+        }.context(FileUriSnafu {
+            input_source: input.input.as_str(),
+        })?;
```
`debug!` is the wrong log level for a failure here. We expect the URI to have parsed properly, so an error here would be `error!` and not `debug!`.

I'd be OK with removing the logging in this case in favor of just always debug logging at the top level as mentioned in the other comment.

anishcheraku replied:
Fixed - updated the `debug!()` statement to be thrown whenever there is an error parsing the Url for all parsed Urls, not just  in the File/HTTP URI cases.

cbgbt commented on sources/api/apiclient/Cargo.toml:
```diff
@@ -44,6 +50,10 @@ tokio-tungstenite = { workspace = true, features = ["connect"] }
 toml.workspace = true
 unindent.workspace = true
 url.workspace = true
+tempfile.workspace = true
```
Should this also be a dev dependency?

anishcheraku replied:
Fixed - It should be under `[dev-dependencies]` since its only used to test the File parser.

cbgbt reviewed (approved):
Very nice work! I have a few comments about logging, but otherwise this LGTM.

anishcheraku commented:
^ Force pushed to fix commit messages and remove unnecessary changes.

KCSesh commented:
With the merge of: https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/714

We also added `base64` uri support. Would be good to add that here

KCSesh committed:
apiclient: apply: consolidate cloud resolvers

Move AWS-specific resolvers (S3, Secrets Manager, SSM) into a separate
cloud_resolvers module with feature-gated CloudError enum. Improve
error handling, documentation, and test coverage.

Signed-off-by: Kyle Sessions <kssessio@amazon.com>

KCSesh committed:
apiclient: apply: add base64 URI resolver support

Signed-off-by: Kyle Sessions <kssessio@amazon.com>

KCSesh committed:
apiclient: enable all URI schemes for network configure

Move SettingsInput and select_resolver to uri_resolver,
so that both apply and network depend on uri_resolver

Signed-off-by: Kyle Sessions <kssessio@amazon.com>

KCSesh committed:
apiclient: separate HTTP and HTTPS resolvers behind TLS feature flag

Split HttpUri into HttpUri (http://) and HttpsUri (https://), gating
HttpsUri behind the tls feature with other TLS-dependent resolvers.

Also adds read_bounded_response helper for size-limited streaming,
improves ARN parsing for all AWS partitions, and relocates TLS tests.

Signed-off-by: Kyle Sessions <kssessio@amazon.com>

KCSesh commented:
Summary of the above ^: 

* Rebased original apiclient apply commits onto develop
* Pinned aws-sdk-s3 to 1.85.0 (1.86+ pulls in crc-fast which doesn't support LTO)
* Updated cached crate to 0.56 to avoid a skip - used by schnauzer
* The existing tls feature flag from [PR #173](https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/173) wasn't enforced around HTTPS, so added TLS feature guards to properly gate
HTTPS, S3, SSM, and Secrets Manager URI resolvers
* Consolidated AWS resolvers into a dedicated tls_resolvers module
* Added base64:// URI resolver support
* Expanded apiclient network configure to support the same URI schemes as apply

bcressey commented on sources/clarify.toml:
```diff
@@ -221,19 +228,15 @@ license-files = [
 license-files = [
-    { path = "LICENSE-APACHE", hash = 0xd0891cfa },
     { path = "LICENSE-MIT", hash = 0xbcce9ab4 },
+    { path = "LICENSE-APACHE", hash = 0xd0891cfa },
```
nit: is this change required? if so why?

KCSesh replied:
Fixed.

bcressey commented on sources/Cargo.toml:
```diff
@@ -121,6 +121,11 @@ aws-lc-rs = "=1.13.0"
+aws-sdk-s3 = "=1.85.0"
+aws-sdk-secretsmanager = "1"
+aws-sdk-ssm = "1"
+aws-smithy-checksums = "=0.63.1"
```
If LTO is causing problems, I'd just remove `-Clto=thin` from `%__global_rustflags_static` via `sed`. That should be the only build using LTO.

KCSesh replied:
Nice! I was avoiding this because I thought it would affect more then just `apiclient`!

bcressey commented on sources/api/apiclient/Cargo.toml:
```diff
@@ -11,15 +11,23 @@ exclude = ["README.md"]
 [features]
 default = ["tls"]
-tls = ["dep:rustls", "dep:aws-lc-rs", "reqwest/rustls-tls-native-roots"]
+tls = ["dep:rustls", "dep:aws-lc-rs", "dep:aws-config", "dep:aws-smithy-types", "dep:aws-sdk-s3", "dep:aws-sdk-secretsmanager", "dep:aws-sdk-ssm", "dep:aws-smithy-checksums", "reqwest/rustls-tls-native-roots"]
```
nit: maybe group these into an `aws-sdk` feature that both `tls` and `fips` require

KCSesh replied:
Added `aws-sdk`, made `tls` depend on it, and I kept fops depending on `tls`.

bcressey commented on sources/api/apiclient/src/network.rs:
```diff
@@ -160,89 +77,23 @@ pub type Result<T> = std::result::Result<T, error::Error>;
 
     #[tokio::test]
     async fn test_get_content_base64() {
-        // Given a valid TOML network configuration encoded in base64
```
Do we need to delete these comments?

KCSesh replied:
Added back - got a bit over zealous on comment deletion based on "whats" but in tests they should stay.

bcressey commented on sources/api/apiclient/src/network.rs:
```diff
@@ -252,44 +103,13 @@ dhcp4 = true";
-        let error_msg = result.unwrap_err().to_string();
-        assert!(error_msg.contains(&format!("Unsupported URI scheme '{expected_scheme}'")));
-        assert!(error_msg.contains("Only file:// and base64: schemes are supported"));
+        assert!(result.unwrap_err().to_string().contains("UTF-8"));
```
Why did this assert get relaxed from `"not valid UTF-8"` to just `"UTF-8"`?

KCSesh replied:
The error now comes from `uri_resolver::Utf8Decode` instead of the old `network::Base64Utf8`, which has a different message format ("Failed to decode as UTF-8 for {uri}"). 

Updated the assertion to match the actual error message, instead of relaxing it to just UTF-8

bcressey commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,590 @@
+use std::path::PathBuf;
+use tokio::io::AsyncReadExt;
+
+pub const MAX_SIZE_BYTES: u64 = 100 * 1024 * 1024;
```
If we run `apiclient apply` in a loop with 100 MiB blobs with a very long `settings.motd`, does anything explode?

KCSesh replied:
So things do explode - at different levels:

1. settings.motd - 64 KB limit (model validation)

2. apiserver JSON payload - 2 MiB limit (actix-web default). Per [PR #2162](https://github.com/actix/actix-web/pull/2162), this appears to be an industry standard with .NET, Tomcat, and PHP using similar values.

3. Datastore partition - scales with OS image size: (os_image_gib * 24) - overhead (~39 MiB for a 2 GiB image). Testing 100 MiB payloads (using many bootstrap-containers to bypass the motd limit) filled the partition with No space left on device.

With this testing done, I think lowering the resolver limit from 100 MiB → 2 MiB to match the API server makes sense, since this is a limit that *technically* already existed. We can increase the limit later if needed, but given the private partition constraints, 2 MiB seems practical for now. I tried to look at the previous discussions to see where 100MiB came from - but didn't fine any conversation on it.

I stress tested - 2 MiB payloads in a loop (200 iterations)  - several times with no issues. 

I would imagine that 2MiB should be plenty of room for users. And worse case, they can apply multiple 2MiB files. I had to create over 200 bootstrap containers with userdata defined to hit the limit. I also created 1 bootstrap container with a huge amount of userdata as well.

bcressey commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,590 @@
+impl UriResolver for FileUri {
+    async fn resolve(&self) -> ResolverResult<String> {
+        use resolver_error::*;
+        tokio::fs::read_to_string(&self.path)
+            .await
+            .context(FileReadSnafu {
+                input_source: self.path.to_string_lossy().into_owned(),
+            })
```
This doesn't enforce the size limitation.

KCSesh replied:
Fixed.

bcressey commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,590 @@
+                max_size: MAX_SIZE_BYTES,
+                uri: self.url.to_string(),
+            }
+        );
```
This implies that a malicious remote could send more than `content_length` bytes and we'd keep reading them all.

Can we instead only read up to `content_length` bytes and assert that we got exactly that many?

KCSesh replied:
Fixed

bcressey commented on sources/api/apiclient/src/tls_resolvers.rs:
```diff
@@ -0,0 +1,550 @@
+                max_size: MAX_SIZE_BYTES,
+                uri: &uri_str,
+            }
+        );
```
Same note here about validating content length. You can look at how `tough` handles this to see if that approach can be reused.

If `reqwest` or the underlying libraries provide guarantees here, then these checks might not be necessary, though it wouldn't hurt to use a reader construct that provides a buffer of a fixed capacity.

KCSesh replied:
Good catch. Looked at how tough handles this in tough/src/io.rs - it uses a [max_size_adapter](https://github.com/awslabs/tough/blob/develop/tough/src/io.rs#L71) that wraps the response stream and errors immediately when cumulative bytes exceed the limit, rather than buffering everything first.

Implemented a `read_bounded_response` helper that:
1. Checks Content-Length header first for early rejection
2. Streams chunks via `resp.chunk()` and tracks cumulative size
3. Errors immediately when limit exceeded - stops reading, doesn't buffer unbounded data

Updated  HTTP(S) resolvers : to use this.

bcressey commented on sources/api/apiclient/src/tls_resolvers.rs:
```diff
@@ -0,0 +1,550 @@
+                input_source: input.to_string(),
+                reason: "must start with 'arn:aws:'".to_string(),
+            }
+        );
```
Why not support ARNs for other partitions?

KCSesh replied:
Added the additional test case to the PR - and I fixed the parsing logic to work on any partition

bcressey commented on sources/api/apiclient/src/tls_resolvers.rs:
```diff
@@ -0,0 +1,550 @@
+    async fn resolve(&self) -> ResolverResult<String> {
+        use tls_resolver_error::*;
+        let cfg = aws_config::load_defaults(aws_config::BehaviorVersion::latest()).await;
+        let client = aws_sdk_s3::Client::new(&cfg);
```
What region does this default to if no region is specified?

KCSesh replied:
`aws_config::load_defaults()` uses the default region provider chain (from [aws-config/src/default_provider/region.rs](https://docs.rs/aws-config/latest/aws_config/default_provider/region/fn.default_provider.html)):

1. Environment variables (`AWS_REGION` / `AWS_DEFAULT_REGION`)
2. Profile file (`~/.aws/config`)
3. EC2 IMDSv2

If none of these provide a region, it returns `None` - there's no hardcoded default.

bcressey commented on sources/api/apiclient/src/tls_resolvers.rs:
```diff
@@ -0,0 +1,550 @@
+
+    #[snafu(display("Invalid SSM URI scheme for '{}', expected ssm://", input_source))]
+    SsmUri { input_source: String },
+
+    #[snafu(display("SSM parameter '{}' did not return a string value", parameter_name))]
+    SsmParameterMissing { parameter_name: String },
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use test_case::test_case;
+
+    #[test_case("https://example.com/foo", "https://example.com/foo"; "https_ok")]
```
Can we add test cases for parsing all the other types?

KCSesh replied:
Those tests were in `uri_resolver` - so they existed - but I kept them there originally with the tls feature flag.

But I have now moved them to `tls_resolver` now to be more clear, and keep things together, as they should have been to begin with.

KCSesh commented:
^ Added:
*  read_bounded_response for proper size limit checks
* Added some default timeout values to all remote URI's (HTTP, HTTPS, S3) (taken from tough) 
  * /// Timeout for establishing a connection.
   pub const CONNECT_TIMEOUT_SECS: u64 = 10;
   * /// Maximum time for an entire request.
    pub const OPERATION_TIMEOUT_SECS: u64 = 30;
* Expanded ARN support to more then commercial partition
* Changed Max Limit from 100MiB to 2 MiB which matches the actix-web default, see [here](https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/554#discussion_r2785209177)
* Fixed some diff deltas (comment removals + reordered clarify.toml) 
* Moved tls unit tests from uri_resolver to tls_resolver (and added a proper stdin check)

KCSesh commented:
^ Nit: Comment fix + removed uneeded helper after refactors

KCSesh commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,517 @@
+//! Defines `UriResolver`, an async trait for fetching UTF-8 text from various URI schemes.
+//! Concrete types parse and resolve a single scheme via `TryFrom` + `resolve()`:
+//!  - `StdinUri` for "-", `FileUri` for `file://`, `HttpUri` for `http://`
```
Nit: Will update to include base64: and also ARN's for SSM and Secrets Manager.

bcressey commented on sources/api/apiclient/src/apply.rs:
```diff
@@ -218,25 +177,100 @@ mod error {
+    #[test_case("ssm://param",                                                      TypeId::of::<crate::uri_resolver::SsmUri>();                    "ssmUri")]
+    #[test_case("arn:aws:ssm:<region>:<account_id>:parameter/<name>",               TypeId::of::<crate::uri_resolver::SsmArn>();                    "ssmArn")]
+    #[test_case("arn:aws:secretsmanager:<region>:<account-id>:secret:<secret-id>",  TypeId::of::<crate::uri_resolver::SecretsManagerArn>();         "secretsmanagerArn")]
+    #[cfg(feature = "tls")]
```
Are both of these `#[cfg(feature = "tls")]` directives needed?

bcressey commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,517 @@
+                size: bytes.len() as u64 + chunk.len() as u64,
+                max_size,
+                uri,
+            }
```
Should avoid the unchecked casts - maybe convert `max_size` into a usize first, e.g.:
```
let max_resp_size = max_size.try_into()?;
while let Some(chunk) = resp.chunk().await.context(HttpBodySnafu { uri })? {
    let resp_size = bytes.len() + chunk.len();
    if resp_size > max_resp_size {
        ...
    }
```

KCSesh replied:
Fixed

bcressey commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,517 @@
+/// Checks Content-Length header first for early rejection, then streams with hard limit.
+pub async fn read_bounded_response(
+    mut resp: reqwest::Response,
+    max_size: u64,
```
nit: not sure we really need this parameter vs. just embedding the constant as we do for timeouts

cbgbt commented on sources/api/apiclient/src/uri_resolver.rs:
```diff
@@ -0,0 +1,517 @@
+    type Error = ResolverError;
+    fn try_from(input: &SettingsInput) -> ResolverResult<Self> {
+        use resolver_error::*;
+        let url = input.parsed_url.clone().context(FileUriSnafu {
```
nit: needless clone

cbgbt commented on sources/api/apiclient/src/tls_resolvers.rs:
```diff
@@ -0,0 +1,635 @@
+}
+
+impl Arn {
+    fn parse(input: &str) -> ResolverResult<Self> {
```
This would technically parse with no error even if there's no region in the ARN, which is actually acceptable for global services but I'm wondering if it could lead to confusing errors here.

KCSesh replied:
Fair - part of my logic was let the SDK handle arn strings after we got the basic verification.. but just adding an empty check makes perfect sense. Fixed 👍

bcressey commented on sources/api/apiclient/src/tls_resolvers.rs:
```diff
@@ -0,0 +1,635 @@
+                key: self.key.clone(),
+            })?;
+
+        let bytes = resp.body.collect().await.context(S3BodySnafu {
```
Maybe something like this:
```
let bytes = resp.body.into_async_read();
let mut cursor = Cursor::new(data);
let mut reader = cursor.take(MAX_SIZE_BYTES);
let mut buffer = Vec::new();
let bytes_read = reader.read_to_end(&mut buffer).await?;
```

KCSesh replied:
I modified it to: `cursor.take(MAX_SIZE_BYTES + 1);` so we can detect when they body is larger than 2MiB and error.

cbgbt commented on sources/api/apiclient/src/tls_resolvers.rs:
```diff
@@ -0,0 +1,635 @@
+        let remainder = uri_str.strip_prefix(PREFIX).context(S3UriSchemeSnafu {
+            input_source: input.input.clone(),
+        })?;
+        let mut parts = remainder.splitn(2, '/');
```
`splitn` returns "at most `n`" values. The  `next()` calls below make sure we get *something* but I think empty strings would parse?

KCSesh replied:
Yeah - I should cover the empty case here too!

cbgbt reviewed (commented):
This LGTM. I think I have a few questions around unusual ARNs - might call for test cases to cover them. I'm otherwise content to approve this PR.

KCSesh committed:
apiclient: apply: add default timeout to requests

Signed-off-by: Kyle Sessions <kssessio@amazon.com>

KCSesh commented:
^ [Diff](https://github.com/bottlerocket-os/bottlerocket-core-kit/compare/f3430c1ab0752d505d8df6c47f8101598a09db87..978e7a6aa018a4352236e2281a3228bd35034fc7)

Addressed the comments: 

* Removed duplicate tls feature flag
* Remove parameter on read_bounded_response
* Added s3 read bound logic 
  * SSM/SecretsManager have built in limits on creation and read
      * SSM (Standard tier: 4 KB max/Advanced tier: 8 KB max)
      * Secrets Manager: 64 KB max per secret value 
* Handled ARN/S3 empty cases directly
* Fix other nits

+ re-ran tests

KCSesh commented:
^ Added additional ARN test cases.

KCSesh commented:
^ Rebased to latest develop

KCSesh commented:
^ [Nit](https://github.com/bottlerocket-os/bottlerocket-core-kit/compare/9d139748a2deb9fee700a1e391a50ed42974457a..2ba547233db39e8b707bbe03009551414200fa44): Added `format_sdk_error` to better produce actual error responses from the SDK making use of [ProvideErrorMetadata](https://docs.rs/aws-smithy-types/1.4.3/aws_smithy_types/error/metadata/trait.ProvideErrorMetadata.html) 

As without this I was just getting `Service error` with no context when buckets didn't exist/params didn't exist, etc. 


i.e 

`
Failed to apply settings: Resolver failed: TLS resolver error: Failed to fetch S3 object 's3://<bucket>/file.toml': NoSuchBucket: The specified bucket does not exist
`
vs 
`
Failed to apply settings: Resolver failed: TLS resolver error: Failed to fetch S3 object 's3://<bucket>/file.toml': service error
`
